compose: Fix mobile keyboard overlap and disable autofill in filter/topic inputs.

hypnoastic opened the pull request:
## What does this PR do?

This PR fixes the mobile keyboard overlap issue where the compose box gets hidden behind the on-screen keyboard when users tap to type on mobile devices.

## Why a JavaScript solution?

CSS-only solutions were evaluated but found insufficient:

- **`dvh` units**: Work for initial layout but do not update dynamically when the keyboard opens/closes on iOS Safari.
- **`env(keyboard-inset-bottom)`**: Limited browser support; does not work reliably across iOS and Android.
- **`interactive-widget=resizes-content` meta tag**: Works on some Android browsers but causes issues on iOS and lacks universal support.

The `visualViewport` API provides the only reliable cross-platform method to detect keyboard height and adjust positioning dynamically.

## Why these specific changes?

The core problem was that on mobile browsers, the compose box (which uses `position: fixed; bottom: 0`) doesn't automatically stay above the keyboard. Different browsers handle this differently:

- **Android Chrome**: Generally resizes the viewport when keyboard opens, so the JavaScript solution using `visualViewport` API worked with minimal tweaking.
- **iOS Safari**: This was much trickier. iOS Safari sometimes resizes the viewport, but other times it just *scrolls* the page instead. This inconsistent behavior meant the compose box would sometimes be above the keyboard and sometimes behind it. We had to use `document.documentElement.clientHeight` instead of `window.innerHeight` to get consistent calculations, and carefully subtract `viewport.offsetTop` to prevent the compose box from scrolling with the chat.

So this wasn't a single issue — it was multiple behaviors across platforms that all needed to be handled:
1. Keyboard detection using `visualViewport` API
2. Handling iOS's scroll-vs-resize inconsistency
3. Keeping compose pinned to visual viewport bottom on scroll
4. Preventing browser autofill popups on channel/topic inputs

## Screenshots

### Android

| Before | | After |
|:------:|:---:|:-----:|
| <img width="200" src="https://github.com/user-attachments/assets/042382b8-c958-4444-a504-10dfc54ddd35" /> <img width="200" src="https://github.com/user-attachments/assets/81da723b-0bb1-4b67-883f-2681c722391c" /> | ➔ | <img width="200" src="https://github.com/user-attachments/assets/0ddeb741-479a-4a2a-a1d0-b98fa6ad9ad1" /> |

### iOS (Tested on Xcode Simulator - iPhone 17 Pro Max)

| Before | | After |
|:------:|:---:|:-----:|
| <img width="200" src="https://github.com/user-attachments/assets/d7e30777-c7b4-4989-ad9c-5a59ad83afbb" /> <img width="200" src="https://github.com/user-attachments/assets/ac108f9e-9e1c-48c8-ab67-17ca5377db82" /> | ➔ | <img width="200" src="https://github.com/user-attachments/assets/ffa8300a-d9fd-4954-82a7-0a032e4621bd" /> |

### iPad (Tested on Xcode Simulator - iPad Pro 11-inch M5)

| After |
|:-----:|
| <img width="500" src="https://github.com/user-attachments/assets/90d96882-3b6e-4b1e-8883-1894081759fb" /> |

<details>
<summary><strong>Self-review checklist</strong></summary>

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability  
  (variable names, code reuse, readability, etc.)
- [x] Followed the [AI use policy](https://zulip.readthedocs.io/en/latest/contributing/contributing.html#ai-use-policy-and-guidelines).
- [x] Communicated decisions, questions, and potential concerns.

### Design & implementation notes
- [x] Explained differences from previous plans (e.g., issue description).
- [x] Highlighted technical choices and bugs encountered.
- [x] Called out remaining decisions and concerns.

### Testing
- [x] Automated tests verify logic where appropriate.

### Commit discipline
- [x] Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).
- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

### Manual review completed
- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions, and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.

</details>

---

Fixes #36676

alya commented:
Thanks for working on this! Please clean up your commit history and post again to request a review. See [here](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html) for guidelines.

Your PR will also need to pass tests.

hypnoastic committed:
compose: Fix mobile keyboard overlap on iOS/Android

The compose box was getting hidden behind the on-screen keyboard
on mobile devices. This adds a keyboard handler using the
visualViewport API to detect keyboard height and adjust compose
position. Also prevents browser autofill on channel/topic inputs.

Fixes: #36676

hypnoastic commented:
Hi, just checking in on this PR — happy to make any changes or split things up if that would help with review. Thanks!

alya commented:
Folks have been taking some time off over the holidays; hopefully @karlstolley will have a chance to take a look soon.

hypnoastic commented:
I’ve resolved the merge conflicts and fixed the keyboard overlapping issue on the latest iOS after testing on the Xcode Simulator. I’ve also updated the PR with a detailed explanation of why a JavaScript-heavy solution is required for this behavior.

One issue still persists: on iOS Safari, the page sometimes scrolls up instead of resizing when the keyboard opens. This appears to be inherent iOS Safari behavior and isn’t something we can fully control from the web layer.

I believe this is separate from the keyboard overlap fix addressed in this PR. If it makes sense, we could track this behaviour in a separate issue. I’m happy to iterate further if there are additional concerns.

<img width="250" height="2868" alt="Simulator Screenshot - iPhone 17 Pro Max - 2026-01-17 at 23 57 46" src="https://github.com/user-attachments/assets/17008307-cd37-401b-adce-64f4a3a827c2" />

hypnoastic commented:
Hi @karlstolley — I just tested the latest `main` on my devices.  
Here are the screenshots for reference

### iOS
<img src="https://github.com/user-attachments/assets/509db694-71ef-4219-b433-7c804a9797d0" width="250" />
<img src="https://github.com/user-attachments/assets/1642b9f9-8ddb-44ca-9927-78dd1152172f" width="250" />
<img src="https://github.com/user-attachments/assets/4f5be5de-9907-4252-b985-5a64a8c80f6c" width="250" />

### Android
<img src="https://github.com/user-attachments/assets/56e8d778-663a-42dc-9de1-c760ae6ea1fd" width="250" />
<img src="https://github.com/user-attachments/assets/b0289fa7-6c26-43bb-8935-0332ed1c14ac" width="250" />
<img src="https://github.com/user-attachments/assets/3340ac41-2087-4b75-a55c-46b91668a1d9" width="250" />

The issues still persist even with the changes made in #37616.
Also, thanks for pointing out , I changed the self-review-checklist to Zulip's own checklist.

hypnoastic commented:
Hi, any updates on this?

hypnoastic commented:
Hi, I am closing this PR as there is no response on this and I don't think this is a ideal and stable solution for this issue.
